charts/redpanda/helpers.go

```diff
@@ -62,8 +62,46 @@ func FullLabels(dot *helmette.Dot) map[string]string {
 		"app.kubernetes.io/component":  Name(dot),
 	}
 
-	// TODO Shouldn't our defaults take precedence over user provided labels?
-	return helmette.Merge(defaults, labels)
+	return helmette.Merge(labels, defaults)
+}
+
+func StatefulSetPodLabelsSelector(dot *helmette.Dot, existingStatefulSetLabelSelector map[string]string) map[string]string {
+	if existingStatefulSetLabelSelector != nil {
+		return existingStatefulSetLabelSelector
+	}
+
+	values := helmette.Unwrap[Values](dot.Values)
+
+	commonLabels := map[string]string{}
+	if values.CommonLabels != nil {
+		commonLabels = values.CommonLabels
+	}
+
+	component := fmt.Sprintf("%s-statefulset",
+		strings.TrimSuffix(helmette.Trunc(51, Name(dot)), "-"))
+
+	defaults := map[string]string{
+		"app.kubernetes.io/component": component,
+		"app.kubernetes.io/instance":  dot.Release.Name,
+		"app.kubernetes.io/name":      Name(dot),
+	}
+
+	return helmette.Merge(commonLabels, defaults)
+}
+
+func StatefulSetPodLabels(dot *helmette.Dot, existingStatefulSetPodTemplateLabels map[string]string) map[string]string {
+	if existingStatefulSetPodTemplateLabels != nil {
+		return existingStatefulSetPodTemplateLabels
+	}
+
+	values := helmette.Unwrap[Values](dot.Values)
+
+	statefulSetLabels := map[string]string{}
+	if values.Statefulset.PodTemplateLabels != nil {
+		statefulSetLabels = values.Statefulset.PodTemplateLabels
+	}
+
+	return helmette.Merge(statefulSetLabels, StatefulSetPodLabelsSelector(dot, nil))
```

Comment:
I need to think about this more but could passing `nil` here result in a broken StatefulSet?

Reply:
I didn't see problem when performing `go test ./... -short`. My helm version

```
$ helm version
version.BuildInfo{Version:"v3.13.2", GitCommit:"2a2fb3b98829f1e0be6fb18af2f6599e0f4e8243", GitTreeState:"clean", GoVersion:"go1.21.4"}
```

Reply:
It seems I have some problem.

```
helm.go:84: [debug] template: redpanda/templates/statefulset.yaml:54:18: executing "redpanda/templates/statefulset.yaml" at <include "statefulset-pod-labels" .>: error calling include: template: redpanda/templates/_statefulset.tpl:33:10: executing "statefulset-pod-labels" at <include "redpanda.StatefulSetPodLabels" (dict "a" (list . $sts))>: error calling include: template: redpanda/templates/_helpers.go.tpl:90:64: executing "redpanda.StatefulSetPodLabels" at <include "redpanda.StatefulSetPodLabelsSelector" (dict "a" (list $dot nil))>: error calling include: template: redpanda/templates/_helpers.go.tpl:59:42: executing "redpanda.StatefulSetPodLabelsSelector" at <dig "spec" "selector" "matchLabels" nil $statefulSet>: error calling dig: interface conversion: interface {} is nil, not map[string]interface {}
```

Reply:
Problem solved and the `chart_test.go` generated new `yaml` golden file `charts/redpanda/testdata/31-overwrite-statefulset-pod-labels-values.yaml.golden`.